wandb/magic_impl.py

```diff
@@ -0,0 +1,354 @@
+    return _magic_defaults
+        
+
+def set_entity(value, env=None):
+    if env is None:
+        env = os.environ
+
+
+def _fit_wrapper(fn, generator=None, *args, **kwargs):
+    trigger.call('on_fit')
+    keras = sys.modules.get("keras")
+    epochs = kwargs.pop("epochs", None)
+
+    magic_epochs = _magic_get_config("magic.keras.fit.epochs", None)
+    if magic_epochs is not None:
+        epochs = magic_epochs
+    callbacks = kwargs.pop("callbacks", [])
+
+    tb_enabled = _magic_get_config("magic.keras.fit.callbacks.tensorboard.enable", None)
+    if tb_enabled:
+        if not any([isinstance(cb, keras.callbacks.TensorBoard) for cb in callbacks]):
+            tb_callback = keras.callbacks.TensorBoard(log_dir=wandb.run.dir)
+            callbacks.append(tb_callback)
+    
+    wandb_enabled = _magic_get_config("magic.keras.fit.callbacks.wandb.enable", None)
+    if wandb_enabled:
+        if not any([isinstance(cb, wandb.keras.WandbCallback) for cb in callbacks]):
+            wandb_callback = wandb.keras.WandbCallback()
+            callbacks.append(wandb_callback)
+
+    kwargs["callbacks"] = callbacks
+    if epochs is not None:
+        kwargs["epochs"] = epochs
+    if generator:
+        return fn(generator, *args, **kwargs)
+    return fn(*args, **kwargs)
+
+
+# NOTE(jhr): need to spell out all useable args so that users who inspect can see args
+def _magic_fit(self,
+        x=None,
+        y=None,
+        batch_size=None,
+        epochs=1,
+        # FIXME: there is more
+        #verbose=1,
+        #callbacks=None,
+        #validation_split=0.,
+        #validation_data=None,
+        #shuffle=True,
+        #class_weight=None,
+        #sample_weight=None,
+        #initial_epoch=0,
+        #steps_per_epoch=None,
+        #validation_steps=None,
+        #validation_freq=1,
+        #max_queue_size=10,
+        #workers=1,
+        #use_multiprocessing=False,
+        *args, **kwargs):
+    return _fit_wrapper(self._fit, x=x, y=y, batch_size=batch_size, epochs=epochs, *args, **kwargs)
+
+
+def _magic_fit_generator(self, generator,
+                    steps_per_epoch=None,
+                    epochs=1,
+                    # FIXME: there is more
+                    #verbose=1,
+                    #verbose=1,
+                    #callbacks=None,
+                    #validation_data=None,
+                    #validation_steps=None,
+                    #validation_freq=1,
+                    #class_weight=None,
+                    #max_queue_size=10,
+                    #workers=1,
+                    ##use_multiprocessing=False,
+                    #shuffle=True,
+                    #initial_epoch=0,
+                    *args, **kwargs):
+    return _fit_wrapper(self._fit_generator, generator=generator, steps_per_epoch=steps_per_epoch, epochs=epochs, *args, **kwargs)
+
+
+def _monkey_keras(keras):
+    keras.engine.Model._fit = keras.engine.Model.fit
```

Comment:
How strict should we be about checking for nested values? Like what if keras.engine is invalid?

Reply:
added a todo, it is already trying to be strict in looking for full module path import string "keras" vs "tensorflow.keras".  I think this hardening will fit well with the next set of changes supporting tf.keras in all its forms tf1 contrib, tf2(beta)